### PR TITLE
capture and display teacher name in registration link #218

### PR DESCRIPTION
**Pull request by farmanahmed888**
This pull request enhances the teacher registration process by requiring and handling a teacher's name throughout the frontend, backend, and database layers. The changes ensure that a name is collected from the user, validated, stored in the database, and included in all relevant API calls and models.

**Frontend changes:**
- Teacher registration form now includes a required "Name" field, and the registration logic passes this value through to the backend. [[1]](diffhunk://#diff-6fd0a583f56e3fad06c1efcb78db0ba545432759844f63414de3cdf275b65394L10-R31) [[2]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1L62-R64) [[3]](diffhunk://#diff-e6df0fa093ee0ba38650ea097cde7960b34907952eaed0ed733b6949f366a9f1L74-R74)

**Backend and API changes:**
- The backend registration endpoint and validation now require the `name` field, returning an error if it is missing.
- The `teacherService.registerTeacher` method and all related backend adapters now accept and store the teacher's name. [[1]](diffhunk://#diff-2e505d92dc212cf1416aab38797fd985e07e953db2b0bde8854a25fdb5e2635dL30-R30) [[2]](diffhunk://#diff-2e505d92dc212cf1416aab38797fd985e07e953db2b0bde8854a25fdb5e2635dR39) [[3]](diffhunk://#diff-0da79c584255ac281126bb0ae003d0655938974c4235f12496e6eb03ceaa5dd0L64-R68) [[4]](diffhunk://#diff-34a1ce7f0627431fc74b7b2bef7e192dc21d27ccdf4907d966407065748897cbL25-R26)

**Database/model changes:**
- The `Teacher` database schema/model now includes a required `name` field.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The main change involves adding a `name` field to the teacher registration process, propagating updates across frontend, backend, and database adapters. The code now collects, validates, and stores teacher names, enhancing data completeness.

### Key Issues & Recommendations

1. **Type & Validation Checks**:
   - The frontend and backend now require `name`, but no validation (e.g., non-empty string) is enforced.
   - Add input validation for `name` on both client and server (e.g., minimum length, character validation).

2. **Consistency & Security**:
   - Ensure password handling (e.g., hashing) occurs before database insertion; no evidence in code.
   - Confirm all pathways handle errors gracefully, especially if `name` is missing or malformed.

3. **Redundancies**:
   - `insertTeacher` functions are duplicated across Firebase and native DB adapters; consider unifying or at least verifying consistent data models.

4. **Naming & Coding Style**:
   - Variable naming for `name` is consistent; ensure all references (UI labels, messages) are updated accordingly.
   - Some error messages still mention only "Phone and password" – update to include "Name" for clarity.

5. **Test Coverage**:
   - No mention of unit/integration tests for the new `name` parameter.
   - Add tests to verify registration with valid/invalid `name`.

### Design & Complexity
- The approach of adding `name` across layers is straightforward, but ensure validation logic is centralized for maintainability.
- There’s no evidence of input sanitization beyond presence checks; consider adding more robust validation.

### Overall Score
**80/100**

- The code changes correctly propagate the new `name` field but require improved validation, error handling, and testing.
- The pattern is standard; avoid duplication and enforce validation standards more strictly.

---

**Recommended Next Steps**:
- Add validation for `name`.
- Ensure password hashing before storage.
- Update tests to cover the `name` parameter.
- Standardize error messages and validation feedback.

**Review by a4i-architect — COMMENTED** [review]
**High-Level Summary:**

The code modifications primarily focus on adding a "name" field to the teacher registration process, propagating it from frontend to backend, database models, and service layers, and enhancing validation and test coverage accordingly.

**Key Observations & Issues:**

- **Code Consistency & Validation:**
  - Proper trimming and validation of the `name` field are added, but only on API level. Consider cross-validating on frontend as well.
  - Length validation (max 200) is handled in API, which is good, but the frontend input doesn't enforce this boundary.
  - Validation for `name`: missing checks for whitespace-only or empty strings in frontend component.
  - The tests for name validation are comprehensive, covering empty, whitespace, length, and duplication.

- **Data Layer & Persistency:**
  - Database schemas and insertion methods correctly include the `name`. No issues here.
  - Migration from native to Firebase/Firestore appears consistent.

- **Design & Architecture:**
  - The propagation pattern (frontend → API → database) is straightforward and well-handled.
  - Validation logic is duplicated between frontend and backend; consider DRYing or central validation schemas (e.g., using Yup or Joi).
  - No significant architectural concerns; the focus on adding a new field is correctly managed across layers.

- **Potential Improvements:**
  - Enforce frontend validation for name length and whitespace early.
  - Use shared validation schemas to prevent inconsistencies.
  - Add unique name constraints if required.
  - Enhance error messages for clarity (e.g., specify "Name is required" vs. just generic messages).

- **Tests:**
  - Well-designed tests covering various invalid inputs.
  - Tests confirm that the name is stored trimmed, validated for length, and required.
  - Good to cover duplicate phone numbers conflict scenario.

**Overall Assessment & Score:**

- **Functionality implementation is correct and consistent across layers.**
- **Tests are comprehensive.**
- **Some frontend validation could be improved for UX.**

**Score: 85/100**

**Acceptance:** Yes, with minor frontend validation enhancements for better user experience.

**Review by a4i-architect — COMMENTED** [review]
### Review Summary

**Key Changes & Intentions:**
- Added `name` field to teacher registration:
  - Frontend: Introduced `teacherName` state and input.
  - Hooks & Service: Pass the name for registration.
  - Backend & DB: Store `name` in Firebase and native MongoDB schemas.
  - Middleware: Validate and trim name, ensure non-empty.
  - Tests: Cover new scenarios, including presence, trimming, duplicate phone, and validation.

**Major Issues & Recommendations:**

1. **Code Quality & Consistency**
   - The validation logic for `name` is thorough but could be centralized. Consider a utility validation function for string fields to avoid repetition.
   
2. **Type Safety & Validation**
   - Frontend: Ensure the `name` input has constraints (e.g., `minLength`, `maxLength`) to prevent unnecessary backend validation.
   - Backend: The validation for `name` is minimal; reinforce with schema validation or Joi.

3. **Database & Schema Changes**
   - Properly handle existing data that lacks `name`. Migration might be needed.
   - Ensure all create/update functions enforce `name` as required, not just during registration.

4. **Testing**
   - The test suite is comprehensive and well-structured.
   - Test for duplicate phone numbers at tenant level is good.
   - Consider testing edge cases like name with only whitespace, special characters, or extremely long strings.

5. **Design & Patterns**
   - The pattern of passing `name` through layers is sound but ensure consistent validation.
   - The approach is straightforward: add field, validate, store, and test.

6. **Potential Improvements**
   - Centralize string validation logic.
   - Add schema validation on the backend (using Joi or similar).
   - Verify that duplicate phone numbers are unique per tenant; currently, uniqueness depends on indexes and duplicate checks.

### Final Score: **85/100**

- Acceptable after minor improvements:
  - Consolidate validation logic.
  - Confirm schema validation consistency.
  - Ensure backward compatibility if existing data lacks `name`.

---

**References:**
- Validation best practices: [Joi Documentation](https://joi.dev/)
- Mongoose schema design: [Mongoose Validation](https://mongoosejs.com/docs/validation.html)

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The introduced modifications focus primarily on adding a `name` field to the teacher registration process, propagating this via frontend, backend, database schema, and tests. Enhancements include input validation, updating API endpoints, models, and service layers to handle `name`. The existing structure remains consistent, but some issues need addressing for cleanliness and robustness:

### Critical Issues & Recommendations

1. **Unnecessary/Incomplete Tests**:
   - Tests checking for empty or whitespace-only `name` are present but lack comprehensive validation. They should explicitly test that such names are rejected with appropriate messages.
   
2. **Validation Logic**:
   - Frontend adds a new input field with basic `useState`. Backend validation now trims and enforces `name` presence, which is good. However, ensure consistent validation (e.g., prevent only whitespace). Currently, you check `typeof name !== "string"` but should also check for empty strings after trimming.

3. **Type Safety & Consistency**:
   - Some backend code assumes `name` is always a string, but extra validation could improve robustness (e.g., non-string or empty after trim). 

4. **Duplicate Code & Overly Complex Validation**:
   - Validation checks should be centralized or use existing validation libraries (like `validator`) for name, in addition to phone and password, for consistency.
   
5. **Database & Model Schema**:
   - The schema updates are correct. Ensure existing data migrations are handled if needed.

6. **API & Service Layers**:
   - Propagation of `name` is consistent and well-handled. However, consider error handling for missing or invalid `name` at API level to provide clearer client feedback.

7. **Testing Coverage**:
   - Tests cover many scenarios, but should add specific tests for:
     - Leading/trailing whitespace trimming
     - Conflicts if a teacher with the same phone exists
     - Validation errors for invalid or missing `name`

8. **Coding Style & Best Practices**:
   - No major issues, but consider consolidating validation logic, removing redundant checks, and ensuring consistent error messages.

### Score: 78/100

**Next steps**:
- Enhance validation (e.g., enforce non-empty after trim, restrict invalid characters if needed).
- Improve test coverage for edge cases.
- Consider refactoring validation logic centrally.
- Validate and handle duplicate phone numbers more explicitly, if not already.

**Reference Articles**:
- Effective validation patterns: [https://express-validator.github.io/](https://express-validator.github.io/)
- Mongoose schema best practices: [https://mongoosejs.com/docs/guide.html](https://mongoosejs.com/docs/guide.html)
- Writing comprehensive API tests: [https://experienceleague.adobe.com/docs/target/using/developer-reference/http-api/validation.html](https://experienceleague.adobe.com/docs/target/using/developer-reference/http-api/validation.html)

File: backend-server/src/auth/teacher/teacherAuthProviderMiddleware.js [hunk 1]
```diff
@@ -47,13 +47,17 @@ module.exports = {
...
+    const { phoneNumber, password, name } = req.body;
     const tenantId = req.userId;
-    if (!phoneNumber || !password || !tenantId) {
+    if (!phoneNumber || !password || !tenantId || typeof name !== "string") {
       return res.status(STATUS.BAD_REQUEST).json({
-        message: "Phone number, password, and tenantName are required",
+        message: "Phone number, password, name, and tenantName are required",
       });
     }
+    const trimmedName = name.trim();
```
Comment (Soumabha) [inline backend-server/src/auth/teacher/teacherAuthProviderMiddleware.js:63]: Add it to the previous conditional validation itself. name.trim() === 0

File: backend-server/tests/teacherRegistration.test.js [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+  beforeEach(async () => {
+    process.env.AUTH_TYPE = "native";
+    const collections = await mongoose.connection.db.collections();
+    for (const collection of collections) {
+      await collection.deleteMany({});
+    }
+  });
+
+  afterEach(async () => {
```
Comment (Soumabha) [inline backend-server/tests/teacherRegistration.test.js]: Considering we are anyway doing the deletion `beforeEach()` why do we need to do `afterEach()`?

File: backend-server/tests/teacherRegistration.test.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    email: TEST_EMAIL,
+    password: "hashedplaceholder",
+    tenantName: TEST_TENANT_NAME,
+  });
+  const token = jwt.sign({ id: tenant._id.toString() });
+  return { tenant, token };
+}
+
+describe("Teacher registration API", () => {
+  beforeAll(async () => {
```
Comment (Soumabha) [inline backend-server/tests/teacherRegistration.test.js]: These should go in as integrationSetup for the test classes exposing methods for setup, teardown and clear database which can be used by the integration test files.

File: backend-server/tests/teacherRegistration.test.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
+process.env.AUTH_TYPE = "native";
+process.env.SECRET_KEY = "test-secret-key-for-testing-purposes-123";
+
+process.env.AZURE_STORAGE_ACCOUNT_NAME = "mockaccountname";
+process.env.AZURE_STORAGE_ACCOUNT_KEY = "mockkeymockkeymockkeymockkeymockkeymockkeymockkeymockkey";
+
```
Comment (Soumabha) [inline backend-server/tests/teacherRegistration.test.js]: Not all tests added here are applicable for integration testing. Any detail which does not require multi class interactions should be written as unit test cases.

Here is a proposed segregation (please validate before segregating yourself)
Image

File: ContentWebApp/src/components/AllContent/RegistrationTab/TeachersList.js [hunk 5]
```diff
@@ -16,7 +16,7 @@ const TeachersList = ({ teachers, selectedTeacherId, onSelectTeacher }) => {
                 String(teacher._id) === String(selectedTeacherId) ? "selected" : ""
               }`}
             >
-              {teacher.phoneNumber}
+              {[teacher.name, teacher.phoneNumber].filter(Boolean).join(" — ")}
```
Comment (Soumabha) [inline ContentWebApp/src/components/AllContent/RegistrationTab/TeachersList.js]: Always show teacher name, default to "unknown" if it is not present in the database.

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The primary change introduces a new `name` field to teachers, with corresponding updates in backend and frontend logic. The code now captures, validates, stores, and displays teacher names, improving data completeness. Tests are added to ensure validation and proper storage.

### Issues & Recommendations

1. **Code Consistency & Validation:**
   - Frontend form strips `name` but backend appends it directly without trimming; ensure backend validation trims and sanitizes (`teacherAuthProviderMiddleware.js`). The backend currently trims only if `name` passes existence checks, but it’s safer to do trimming explicitly.
   - Improve password validation to enforce password strength if not already done (not visible in code).

2. **Type & Validation Checks:**
   - The frontend already checks `name` as a string. Backend validation checks `typeof name` and `name.trim()`. Ensure backend rejects strings with only whitespace.
   - In `Teacher.js`, name is marked `required: true`, which is good.

3. **Duplicate & Dead Code:**
   - The test code for registration validation is solid. No duplicated validation logic is evident.
   - `registerTeacher` in `teacherService.js` and the backend database adapters correctly now handle `name`.

4. **Design Considerations:**
   - Exposing `name` in all teacher fetches (list, detailed endpoints) is consistent and good.
   - Consider adding a display name fallback (`"unknown"`) if `name` is missing, which is already handled in `TeachersList.js`.

5. **Other Observations:**
   - `.gitignore` updates seem unrelated; consider removing unrelated tracking if not relevant.
   - No major architectural issues identified; the change is straightforward.

### Scoring & Final Notes

- **Score: 80/100**  
  The change is well-implemented but should include strict trimming and validation in backend middleware. Additional password policy validation and sanitation can improve security.

- **Acceptance Recommendation:**  
  Accept with recommendations to improve server-side validation consistency and possibly enhance password policies.

**Relevant Articles:**
- [Effective validation patterns](https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation)
- [Designing API versioning & validation](https://restfulapi.net/resource-naming/)

---

**Summary:**  
Solid extension improving data quality; minor backend validation robustness needed.

File: ContentWebApp/src/components/AllContent/RegistrationTab/TeachersList.js (hunk 5 above)
Comment (farmanahmed888) [reply ContentWebApp/src/components/AllContent/RegistrationTab/TeachersList.js]: done

**Review by theainerd — APPROVED** [review]
**Verdict**: Approved (with nits)

The implementation looks good and handles the requirements.

One minor nit: The error message in `teacherAuthProviderMiddleware.js` mentions `tenantName` but the code checks for `tenantId`. This is pre-existing but worth cleaning up.

LGTM!
